controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+	vm *vmopv1.VirtualMachine) error {
+
+	_, err := controllerutil.CreateOrPatch(ctx, r.Client, vm, func() error {
+		return controllerutil.SetControllerReference(rs, vm, r.Scheme())
+	})
+
+	return err
+}
+
+func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineReplicaSetContext) (ctrl.Result, error) {
 	if !controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
-		// Set the finalizer and return immediately so the object is patched immediately.
+		// Set the finalizer and return so the object is patched immediately.
 		controllerutil.AddFinalizer(ctx.ReplicaSet, finalizerName)
-		return nil
+		return ctrl.Result{}, nil
 	}
 
 	ctx.Logger.Info("Reconciling VirtualMachineReplicaSet")
 
-	return nil
+	if ctx.ReplicaSet.Labels == nil {
+		ctx.ReplicaSet.Labels = make(map[string]string)
+	}
+
+	if ctx.ReplicaSet.Spec.Selector.MatchLabels == nil {
+		ctx.ReplicaSet.Spec.Selector.MatchLabels = make(map[string]string)
+	}
+
+	if ctx.ReplicaSet.Spec.Template.Labels == nil {
+		ctx.ReplicaSet.Spec.Template.Labels = make(map[string]string)
+	}
+
+	selectorMap, err := metav1.LabelSelectorAsMap(ctx.ReplicaSet.Spec.Selector)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err,
+			"failed to convert VirtualMachineReplicaSet %q label selector to a map",
+			ctx.ReplicaSet.Name)
+	}
+
+	// Get all VirtualMachines linked to this VirtualMachineReplicaSet.
+	allVMs := &vmopv1.VirtualMachineList{}
+	if err := r.Client.List(ctx,
+		allVMs,
+		client.InNamespace(ctx.ReplicaSet.Namespace),
+		client.MatchingLabels(selectorMap),
+	); err != nil {
+		return ctrl.Result{},
+			fmt.Errorf("failed to list virtual machines matched by the VirtualMachineReplicaSet. selector: %q, err: %v",
+				selectorMap, err)
+	}
+
+	// Filter out irrelevant VirtualMachines (i.e. IsControlledBy something else) and
+	// claim orphaned machines.
+	// Note that claim does not mean that we will try to conform the VM to meet the
+	// desired spec since that might involve moving the VM across
+	// compute and storage boundaries.
+	// VirtualMachines in deleted state are not excluded.
+	filteredVMs := make([]*vmopv1.VirtualMachine, 0, len(allVMs.Items))
+	for idx := range allVMs.Items {
+		vm := &allVMs.Items[idx]
+		// Skip this VirtualMachine if it is controlled by someone else.
+		if shouldExcludeVirtualMachine(ctx.ReplicaSet, vm) {
+			continue
+		}
+
+		// Attempt to adopt VirtualMachine if it has no controller references.
+		if metav1.GetControllerOf(vm) == nil {
+			if err := r.adoptOrphan(ctx, ctx.ReplicaSet, vm); err != nil {
+				ctx.Logger.Error(err, "VirtualMachineReplicaSet failed to adopt VirtualMachine", "vm", vm.Name)
+				r.Recorder.Warnf(ctx.ReplicaSet, "FailedAdopt", "Failed to adopt VirtualMachine %q: %v", vm.Name, err)
+
+				continue
+			}
+			ctx.Logger.Info("Adopted VirtualMachine", "vm", vm.Name)
+			r.Recorder.Eventf(ctx.ReplicaSet, "SuccessfulAdopt", "Adopted VirtualMachine %q", vm.Name)
+		}
+
+		filteredVMs = append(filteredVMs, vm)
```

Comment:
If the VM has a _different_ controller owner do we really want to add the label to it below?

Reply:
Oh I missed what `shouldExcludeVirtualMachine()` does. But maybe just inline that function so we don't have two calls to `metav1.GetControllerOf()` right after eachother?

Reply:
Fixed.